services/api/service.go

```diff
@@ -381,6 +398,80 @@ func (api *RelayAPI) startValidatorRegistrationDBProcessor() {
 	}
 }
 
+// simulateBlock sends a request for a block simulation to blockSimRateLimiter.
+func (api *RelayAPI) simulateBlock(opts blockSimOptions) error {
+	t := time.Now()
+	simErr := api.blockSimRateLimiter.send(opts.ctx, opts.req, opts.highPrio)
+	log := opts.log.WithFields(logrus.Fields{
+		"duration":   time.Since(t).Seconds(),
+		"numWaiting": api.blockSimRateLimiter.currentCounter(),
+	})
+	if simErr != nil {
+		log.WithError(simErr).Error("block validation failed")
+	} else {
+		log.Info("block validation successful")
+	}
+	return simErr
+}
+
+// Demotes a set of builders who have matching collateral IDs. This function
+// continues to try demoting even if an error is encountered.
+func (api *RelayAPI) demoteBuildersCollateralID(builderPubkey string) {
+	collateralPubkeys := []string{builderPubkey}
+	// Fetch builder collateral_id.
+	builder, err := api.db.GetBlockBuilderByPubkey(builderPubkey)
+	if err != nil {
+		api.log.WithError(err).Error("unable to get builder from database")
+	}
+	collateralID := builder.CollateralID
+
+	// Fetch additional builder pubkeys using the collateral_id.
+	otherBuilders, err := api.db.GetBlockBuildersFromCollateralID(collateralID)
+	if err != nil {
+		api.log.WithError(err).Error("unable to get extra builders from collateral id")
+	}
+	for _, b := range otherBuilders {
+		collateralPubkeys = append(collateralPubkeys, b.BuilderPubkey)
```

Comment:
It may be cleaner/shorter/faster to do `collateralPubkeys = append(otherBuilders, b.BuilderPubkey)`.

Reply:
oh I realized that `GetBlockBuildersFromCollateralID` returns all the builders, so we can just use that slice rather than trying to append anything. this addresses the idempontency comment above too.